pkg/kubelet/cm/topologymanager/numatopology/numa_topology.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package numatopology
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/bitmask"
+)
+
+type Topology struct {
+	numaNodes []int
+	distances [][]uint64
+}
```

Comment:
Something hasn't been sitting right with me about this abstraction, and I think it's because I'd rather see it implemented as below:
```
type Distances [][]uint64

type Topology struct {
    Nodes []int
    Distances Distances
}

func New(nodes []cadvisorapi.Node) (*Topology, error) {
    ...
}

func (d Distances) CalculateAverageFor(bm bitmask.BitMask) float64 {
    ...
}
```

The main things to note being:
1. There is no `SysFs` abstraction creeping into the API. Once we have `cadvisor` support only this package should have to change -- not the code calling this package.
2. The caller has direct access to the `Topology.Nodes` and `Topology.Distances` fields so there is no need for any `Getter` functions.

For testing (which is where I see you actually caring about the `SysFs` type, you can introduce a hidden helper function if you like called `new()` that *does* take a `SysFs` and call it from both your test code and the top-level `New()` call (which would just hard code passing in `realSysFs`).

Reply:
Actually thinking about this even more overnight, I don't think we should have a subpackage for `numatopology` at all. Instead, I think the abstraction should be called `NUMAInfo`, it should live directly in the `topologymanager` package, and it should directly provide the function needed to calculate the "fitness" score. We can then pass this to the call to `compareHints()` to give us both the set of numa nodes (for calucalting the default mask) as well as directly call the fitness function without passing an extra function pointer.

Reply:
It makes sense to remove this subpackage. One thing I am wondering if we just pass `NUMAInfo` to `mergeFilteredHints`:
```go
func mergeFilteredHints(filteredHints [][]TopologyHint, numaInfo NUMAInfo) TopologyHint {
	// Set bestNonPreferredAffinityCount to help decide which affinity mask is
	// preferred amongst all non-preferred hints. We calculate this value as
	// the maximum of the minimum affinity counts supplied for any given hint
	// provider. In other words, prefer a hint that has an affinity mask that
	// includes all of the NUMA nodes from the provider that requires the most
	// NUMA nodes to satisfy its allocation.
	bestNonPreferredAffinityCount := maxOfMinAffinityCounts(filteredHints)
	defaultAffinity := numaInfo.DefaultAffinityMask()

	var bestHint *TopologyHint
	iterateAllProviderTopologyHints(filteredHints, func(permutation []TopologyHint) {
		// Get the NUMANodeAffinity from each hint in the permutation and see if any
		// of them encode unpreferred allocations.
		mergedHint := mergePermutation(defaultAffinity, permutation)

		// Compare the current bestHint with the candidate mergedHint and
		// update bestHint if appropriate.
		bestHint = compareHints(bestNonPreferredAffinityCount, bestHint, &mergedHint, numaInfo)
	})

	if bestHint == nil {
		bestHint = &TopologyHint{defaultAffinity, false}
	}

	return *bestHint
}
```

and to `compareHints`:
```go
func compareHints(bestNonPreferredAffinityCount int, current *TopologyHint, candidate *TopologyHint, numaInfo NUMAInfo) *TopologyHint {
// implementation
}
```

how should I determine which "fitness" function to use? If I would add it as a function to `NUMAInfo` then I would have to put a logic to determine if it should be `preferNarrowerHintFunc` or `preferClosestNUMAFunc` inside `NUMAInfo` constructor. Something like:
```go
func NewNUMAInfo(nodes []cadvisorapi.Node, topologyPolicyName string, opts PolicyOptions) *NUMAInfo{
// ...
    fitnessFunction := preferNarrowerHintFunc
    if (topologyPolicyName == PolicyBestEffort || topologyPolicyName ==PolicyRestricted) && opts.PreferClosestNUMA {
        fitnessFunction = p.preferClosestNUMAFunc
    }
}

return &NUMAInfo{FitnessFunc: fitnessFunction}
}
```

or should I also policyOptions to `compareHints` ?

Reply:
Or we can just make `NewNUMAInfo` returns default `preferNarrowerHintFunc` as `FitnessFunc` and overwriting this function will be moved to `NewBestEffortPolicy` and `NewRestrictedPolicy` based on `PolicyOpts` something like:
```go
// NewBestEffortPolicy returns best-effort policy.
func NewBestEffortPolicy(numaInfo *NUMAInfo, topologyPolicyOptions map[string]string) (Policy, error) {
	return &bestEffortPolicy{numaNodes: numaNodes}
	opts, err := NewPolicyOptions(topologyPolicyOptions)
	if err != nil {
		return nil, err
	}
        if opts.PreferClosestNuma {
            numaInfo.FitnessFunc = preferClosestNUMAFunc
        }
	return &bestEffortPolicy{numaInfo:numaInfo, opts: opts}, nil
}
```

Reply:
What about defining `NUMAInfo` as:
```
type NUMAInfo struct {
    Nodes []int
    Distances map[int]map[int]int
}

func NewNUMAInfo(nodes []cadvisorapi.Node) *NUMAInfo {
...
}

func (n *NUMAInfo) Narrowest(m1 bitmask.Bitmask, m2 bitmask.Bitmask) bitmask.Bitmask {
...
}

func (n *NUMAInfo) Closest(m1 bitmask.Bitmask, m2 bitmask.Bitmask) bitmask.Bitmask {
...
}
```

And then another abstraction called HintMerger defined as:
```
type HintMerger struct {
    NUMAInfo *NUMAInfo
    FilteredHints [][]TopologyHint
    BestNonPreferredAffinityCount int,
    CompareNUMAAffinityMasks func(m1 bitmask.Bitmask, m2 bitmask.Bitmask) bitmask.Bitmask
}

func NewHintMerger(numaInfo, providerHints, policyName, policyOptions) {
    filteredHints := filterProvidersHints(providerHints)

    merger := HintMerger{
        NUMAInfo: numaInfo,
        FilteredHints: filteredHints,
        BestNonPreferredAffinityCount: maxOfMinAffinityCounts(filteredHints),
        CompareNUMAAffinityMasks: numaInfo.Narrower
    }

    if (topologyPolicyName != PolicySingleNUMA) && opts.PreferClosestNUMA {
        merger.CompareNUMAAffinityMasks = numaInfo.Closest
    }

    return merger
}
```

Then the functions `compareHints`, and `mergeFilteredHints` can hang off of this type with a much smaller set of parameters needing to be passed between them.
```
func (m HintMerger) compareHints(current *TopologyHint, candidate *TopologyHint) *TopologyHint {
...
}

func (m HintMerger) Merge() TopologyHint {
...
}
```

Reply:
That looks much better than my proposal. Let me implement that.

Reply:
I know I am still missing unit tests but can you take a look at `HintMerger`?